# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose
0,1,-8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875.639453,покупка жилья
1,1,-4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080.014102,приобретение автомобиля
2,0,-5623.422610,33,Среднее,1,женат / замужем,0,M,сотрудник,0,145885.952297,покупка жилья
3,3,-4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628.550329,дополнительное образование
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616.077870,сыграть свадьбу
5,0,-926.185831,27,высшее,0,гражданский брак,1,M,компаньон,0,255763.565419,покупка жилья
6,0,-2879.202052,43,высшее,0,женат / замужем,0,F,компаньон,0,240525.971920,операции с жильем
7,0,-152.779569,50,СРЕДНЕЕ,1,женат / замужем,0,M,сотрудник,0,135823.934197,образование
8,2,-6929.865299,35,ВЫСШЕЕ,0,гражданский брак,1,F,сотрудник,0,95856.832424,на проведение свадьбы
9,0,-2188.756445,41,среднее,1,женат / замужем,0,M,сотрудник,0,144425.938277,покупка жилья для семьи


**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Шаг 3. Исследуйте данные и ответьте на вопросы

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_group = data.groupby('children').agg({'debt': ['count','sum']})# сгруппировали данные по столбцу children, для столбца debt применили функции count и sum
data_group

debt      
          count   sum
children             
0         14091  1063
1          4808   444
2          2052   194
3           330    27
4            41     4
5             9     0

выяснили:
- количество клиентов по группировке кол-ву детей с учетом показателя 'debt' - столбец count.
- количество клиентов, имеющих задолженность - столбец sum.

In [23]:
data_group['daily_debt'] = data_group['debt']['sum'] / data_group['debt']['count'] * 100 # разделили сумму должников на общее кол-во клиентов каждой категории кол-ва детей в семье
data_group

debt       daily_debt
          count   sum           
children                        
0         14091  1063   7.543822
1          4808   444   9.234609
2          2052   194   9.454191
3           330    27   8.181818
4            41     4   9.756098
5             9     0   0.000000

- посчитали долю должников в процентах - столбец daily_debt.

In [24]:
data_group[data_group['debt']['count'] >= 300].sort_values(by='daily_debt') # удалили выделяющиеся значения и отсортировали по возрастанию данные по долевому показателю

debt       daily_debt
          count   sum           
children                        
0         14091  1063   7.543822
3           330    27   8.181818
1          4808   444   9.234609
2          2052   194   9.454191

- убрали выделяющиеся значения - значения для клиентов с 4-мя и 5-ю детьми.
- отсортировали таблицу(массив) по столбцу daily_debt по возрастанию.

**Вывод:** 
В данной выборке есть значения, которые сильно выделяются среди остальных - количество клиентов с 4-мя и 5-ю детьми - минимальны. Такие данные искажают результат, поэтому удалили их из таблицы.
Клиенты без детей более надежные заемщики. Клиенты с детьми, а в основной массе те, кто с 1-им или 2-мя детьми, проблемные заемщики.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
def debt_categ (row): #создали новую функцию, которая принимает строку датафрейма как аргумент row
    if row == 1:
        return '1'
    if row == 0:
        return '0'
data['debt_categ'] = data['debt'].apply(debt_categ)
data['debt_categ']

0        0
1        0
2        0
3        0
4        0
        ..
21520    0
21521    0
21522    1
21523    1
21524    0
Name: debt_categ, Length: 21331, dtype: object

- создали столбец debt_categ аналогичный debt для создания сводной таблицы

In [26]:
data_pivot_family = data.pivot_table(index='family_status', columns='debt_categ', values='debt', aggfunc='count')#создали сводную таблицу
data_pivot_family

debt_categ,0,1
family_status,,
Не женат / не замужем,2523,273
в разводе,1105,84
вдовец / вдова,888,63
гражданский брак,3749,385
женат / замужем,11334,927


- создали сводную таблицу с данными по количеству клиентов с задолженностью и без задолженности по семейному положению

In [27]:
data_pivot_family['rate'] = data_pivot_family['1'] / (data_pivot_family['1'] + data_pivot_family['0']) * 100 # создали новый столбец, показывающий долю должников по семейному положению
data_pivot_family

debt_categ,0,1,rate
family_status,,,
Не женат / не замужем,2523,273,9.763948
в разводе,1105,84,7.064760
вдовец / вдова,888,63,6.624606
гражданский брак,3749,385,9.313014
женат / замужем,11334,927,7.560558


- посчитали долю должников в процентах и добавили столбец в сводную таблицу data_pivot_family

In [28]:
data_pivot_family.sort_values(by='rate') # отсортировали по возрастанию таблицу по долевому показателю

debt_categ,0,1,rate
family_status,,,
вдовец / вдова,888,63,6.624606
в разводе,1105,84,7.064760
женат / замужем,11334,927,7.560558
гражданский брак,3749,385,9.313014
Не женат / не замужем,2523,273,9.763948


- отсортировали таблицу(массив) по столбцу rate по возрастанию.

**Вывод:** 
Самый высокий процент должников приходится на людей, не состоящих в отношениях. Наблюдается прямая зависимость наличия у заемщика ярко выраженного чувства ответственности, если заемщик был или находится в зарегистрированном браке. Люди с опытом более организованные/системные - надежные заемщики.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [29]:
data_pivot_income = data.pivot_table(index='total_income_category', columns='debt_categ', values='debt', aggfunc='count')#создали сводную таблицу
data_pivot_income

debt_categ,0,1
total_income_category,,
A,23,2
B,4660,354
C,14568,1353
D,328,21
E,20,2


- создали сводную таблицу с данными по количеству клиентов с задолженностью и без задолженности по категориям с уровнем дохода

In [30]:
data_pivot_income['rate'] = data_pivot_income['1'] / (data_pivot_income['1'] + data_pivot_income['0']) * 100 # создали новый столбец, показывающий долю должников по категории доходов
data_pivot_income

debt_categ,0,1,rate
total_income_category,,,
A,23,2,8.000000
B,4660,354,7.060231
C,14568,1353,8.498210
D,328,21,6.017192
E,20,2,9.090909


- посчитали долю должников в процентах и добавили столбец в сводную таблицу data_pivot_income

In [31]:
data_pivot_income[data_pivot_income['1'] + data_pivot_income['0'] >= 5000].sort_values(by='rate') # отсортировали по возрастанию таблицу по долевому показателю

debt_categ,0,1,rate
total_income_category,,,
B,4660,354,7.060231
C,14568,1353,8.498210


    1. удалили клиентов категории 'A', <b>'D'</b>, 'E', с уровнем дохода до 30 000 рублей, до 50 000 рублей и от 1 млн.рублей соответсвенно, так как таких клиентов меньше всего.
    2. отсортировали таблицу(массив) по столбцу rate по возрастанию.

**Вывод:** 
Клиенты категории 'B' с уровнем дохода 200 000 - 1 млн.руб. надежные заемщики. 

### Как разные цели кредита влияют на его возврат в срок?

In [32]:
data_pivot_purpose = data.pivot_table(index='purpose_category', columns='debt_categ', values='debt', aggfunc='count')#создали сводную таблицу
data_pivot_purpose

debt_categ,0,1
purpose_category,,
операции с автомобилем,3879,400
операции с недвижимостью,9971,780
получение образования,3619,369
проведение свадьбы,2130,183


- создали сводную таблицу с данными по количеству клиентов с задолженностью и без задолженности по категориям цели получения кредита

In [33]:
data_pivot_purpose['rate'] = data_pivot_purpose['1'] / (data_pivot_purpose['1'] + data_pivot_purpose['0']) * 100 # создали новый столбец, показывающий долю должников по категории целей кредита
data_pivot_purpose

debt_categ,0,1,rate
purpose_category,,,
операции с автомобилем,3879,400,9.347978
операции с недвижимостью,9971,780,7.255139
получение образования,3619,369,9.252758
проведение свадьбы,2130,183,7.911803


- посчитали долю должников в процентах и добавили столбец в сводную таблицу data_pivot_purpose

In [34]:
data_pivot_purpose.sort_values(by='rate') # отсортировали по возрастанию таблицу по долевому показателю

debt_categ,0,1,rate
purpose_category,,,
операции с недвижимостью,9971,780,7.255139
проведение свадьбы,2130,183,7.911803
получение образования,3619,369,9.252758
операции с автомобилем,3879,400,9.347978


- отсортировали таблицу(массив) по столбцу rate по возрастанию.

**Вывод:** 
Максимальное количество клиентов те, которые обращаются в банк с целью получения кредита на недвижимость, они же более надежные заемщики. А вот клиентов с целью получения кредита на автомобиль в 2 раза меньше, но должников больше - ненадежные заемщики.

### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Технический сбой(например при выгрузке данных) и человеческий фактор(например, намеренно не внесли данные или забыли внести данные).

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Потому что некоторые значения в столбце/строчке могут сильно выделяться среди остальных значений. Среднеарифметическое значение будет искажть действительность. Поэтому в таких случаях лучше заполнять пропуски медианным значением.

## Шаг 4: общий вывод.

В целях усовершенствования модели кредитного скоринга были проанализированы массивы данных задолженностей по платежам клиентов в зависимости от кол-ва детей, семейного положения, уровня дохода и цели приобретения кредита.
    
На этапе предобработки данных были найдены и заполнены медианным значением пропуски, устранены артефакты и дубликаты.
    
Установлены следующие зависимости:
1.	Кол-во клиентов без детей максимально и процент задолженностей у данной категории минимальный - 7,54%. А вот семья с 2-мя детьми показывают самый высокий процент ненадежности заемщика - 9.45%.  
Кол-во семей с 4-мя и 5-ю детьми значительно меньше, что приводит к удалению их из числа анализа.
2.	Самая рискованная категория клиентов - люди, не состоящие в отношениях - 9,76%. Самые надежные заемщики — это вдовцы/вдовы - 6,62%.
3.	Максимальное кол-во клиентов с доходом 50-200 тыс.руб, но являются ненадежными заемщиками - 8,49%. Добросовестными заемщиками являются клиенты категории 'B' с уровнем дохода 200 000 – 1 млн. руб. 
Клиент заработной платой ниже 50 тыс.руб. и с более 1 млн.руб. редко обращаются в банк - их кол-во минимально относительно других категорий, что приводит к удалению их из числа анализа.
4.	Самая благоприятная для банка цель получения кредита - операции с недвижимостью. Их количество максимально, а процент задолженностей минимальный - 7,25%. А вот клиент с целью получения кредита на операции с автомобилем рискованный заемщик - их кол-во в 2 раза меньше, а процент задолженностей высокий - 9,34.

Результаты исследования показали следующее:
1.	На доход клиентов влияет тип занятости и стаж работы.
2.	Критерии хороших потенциальных заемщиков:
    -	без детей;
    -	состоят или были в браке;
    -	с уровнем дохода 200 000 – 1 млн. руб. - категория B;
    -	с целью получения кредита на недвижимость.

Рекомендации для улучшения системы кредитного скоринга:
Проанализировать зависимости между типом занятости, возрастом заемщика, статуса образования и возвратом кредита в срок.
Дополнить массивы анализом по кол-ву дней заложенности и суммам задолженности, чтобы определить, какие категории входят в серьезную группу риска. 

Так же выявлено много аномальных значений в столбце ‘days_employed, поэтому при сборе данных рекомендуется обратить на это внимание и автоматизировать систему выгрузки, например, с государственной системы – запрос в госуслуги о трудовом стаже клиента. 